Pinta.Core/Effects/ColorGradient.cs

```diff
@@ -0,0 +1,94 @@
+		double minPosition,
+		double maxPosition,
+		IEnumerable<KeyValuePair<double, ColorBgra>> gradientStops)
+	{
+		if (minPosition >= maxPosition) throw new ArgumentException ($"{nameof (minPosition)} has to be lower than {nameof (maxPosition)}");
+		var sortedStops = gradientStops.Select (kvp => new GradientStop (kvp.Key, kvp.Value)).OrderBy (stop => stop.Position).ToArray ();
+		if (sortedStops.Length > 0 && sortedStops[0].Position <= minPosition) throw new ArgumentException ($"Lowest key in {nameof (gradientStops)} has to be greater than {nameof (minPosition)}");
+		if (sortedStops.Length > 0 && sortedStops[^1].Position >= maxPosition) throw new ArgumentException ($"Greatest key in {nameof (gradientStops)} has to be lower than {nameof (maxPosition)}");
+		var distinctPositions = sortedStops.GroupBy (s => s.Position).Count ();
+		if (distinctPositions != sortedStops.Length) throw new ArgumentException ("Cannot have more than one stop in the same position");
+
+		StartColor = startColor;
+		EndColor = endColor;
+		MinimumPosition = minPosition;
+		MaximumPosition = maxPosition;
+		sorted_stops = sortedStops;
+	}
+
+	public sealed override ColorBgra GetColor (double position)
+	{
+		if (position == MinimumPosition) return StartColor;
+		if (position == MaximumPosition) return EndColor;
+		if (!IsMapped (position)) throw new ArgumentOutOfRangeException (nameof (position));
+		if (sorted_stops.Count == 0) return HandleNoStops (position);
+		return HandleWithStops (position);
+	}
+
+	private ColorBgra HandleNoStops (double position)
+	{
+		double valueSpan = MaximumPosition - MinimumPosition;
+		double positionOffset = position - MinimumPosition;
+		double fraction = positionOffset / valueSpan;
+		return ColorBgra.Lerp (StartColor, EndColor, fraction);
+	}
+
+	private ColorBgra HandleWithStops (double position)
+	{
+		int immediateLowerIndex = BinarySearchLowerOrEqual (sorted_stops, position);
+		var immediatelyLower = immediateLowerIndex < 0 ? new GradientStop (MinimumPosition, StartColor) : sorted_stops[immediateLowerIndex];
+		if (immediatelyLower.Position == position) return immediatelyLower.Color;
+		int immediatelyHigherIndex = immediateLowerIndex + 1;
+		var immediatelyHigher = (immediatelyHigherIndex < sorted_stops.Count) ? sorted_stops[immediatelyHigherIndex] : new GradientStop (MaximumPosition, EndColor);
+		double valueSpan = immediatelyHigher.Position - immediatelyLower.Position;
+		double positionOffset = position - immediatelyLower.Position;
+		double fraction = positionOffset / valueSpan;
+		return ColorBgra.Lerp (immediatelyLower.Color, immediatelyHigher.Color, fraction);
+	}
+
+	private readonly record struct GradientStop (double Position, ColorBgra Color);
+
+	/// <returns>Index of number immediately lower than target. If not found, it returns -1</returns>
+	/// <param name="arr">Array assumed to be sorted by key</param>
+	private static int BinarySearchLowerOrEqual (IReadOnlyList<GradientStop> arr, double target)
```

Comment:
Would https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.binarysearch?view=net-7.0 be usable here?

Reply:
The method could be modified to use what you are mentioning, but a simple call to it wouldn't be sufficient, because _exact_ matches are not sufficient. We should be able to search for the _closest_ match in either direction.

Quoting the documentation:

---

Returns
[Int32](https://learn.microsoft.com/en-us/dotnet/api/system.int32?view=net-7.0)

The zero-based index of item in the sorted [List<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-7.0), if item is found; otherwise, a negative number that is the bitwise complement of the index of the next element that is larger than item or, if there is no larger element, the bitwise complement of [Count](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.count?view=net-7.0#system-collections-generic-list-1-count).

---

I would still have to test if the returned index is in range or if it's a bitwise complement of the next element that is larger, and so on, so it would result in several lines of code one way or the other.

Reply:
Yeah it's definitely not just a one-liner, but IMO building on top of the standard library algorithms is better vs rolling our own similar code

Reply:
I just changed the binary search so that it uses standard library methods, too